src/ivc/step_folding_circuit.rs

```diff
@@ -0,0 +1,401 @@
+                    .unwrap();
+
+                Ok((next_step_i, new_U, assigned_input))
+            },
+        )?;
+
+        let z_output =
+            self.step_circuit
+                .synthesize_step(config.step_config, layouter, &assigned_input)?;
+
+        let output_hash = layouter.assign_region(
+            || "generate input",
+            |region| {
+                let mut ctx = RegionCtx::new(region, 0);
+
+                let bits = RO::new(
+                    config.main_gate_config.clone(),
+                    self.input.step_pp.ro_constant.clone(),
+                )
+                .absorb_point(WrapValue::from_assigned_point(
+                    &assigned_input_witness.public_params_hash,
+                ))
+                .absorb_base(WrapValue::Assigned(assigned_next_step_i.clone()))
+                .absorb_iter(assigned_z_0.iter())
+                .absorb_iter(z_output.iter().cloned())
+                .absorb_iter(assigned_new_U.iter_wrap_values())
+                .squeeze_n_bits(&mut ctx, NUM_CHALLENGE_BITS)?;
+
+                MainGate::new(config.main_gate_config.clone()).le_bits_to_num(&mut ctx, &bits)
+            },
+        )?;
+
+        layouter.constrain_instance(
+            assigned_input_witness.input_instance[1].0.cell(),
+            config.instance,
+            0,
+        )?;
+        layouter.constrain_instance(output_hash.cell(), config.instance, 1)?;
+
+        Ok(z_output)
+    }
+}
+
+impl<'link, const ARITY: usize, C, SC, RO, const T: usize>
+    StepFoldingCircuit<'link, ARITY, C, SC, RO, T>
+where
+    C: CurveAffine,
+    C::Base: ff::PrimeFieldBits + ff::FromUniformBytes<64>,
+    C::Scalar: ff::PrimeFieldBits + ff::FromUniformBytes<64>,
+    SC: StepCircuit<ARITY, C::Base> + Sized,
+    RO: ROCircuitTrait<C::Base, Config = MainGateConfig<T>>,
+{
+    pub fn synthesize(
+        &self,
+        config: StepConfig<ARITY, C::Base, SC, T>,
+        layouter: &mut impl Layouter<C::Base>,
+    ) -> Result<[C::Base; ARITY], SynthesisError> {
+        let assigned_z_i: [_; ARITY] = layouter.assign_region(
+            || "assigned_zi",
+            |region| {
+                let mut region = RegionCtx::new(region, 0);
+
+                config
+                    .main_gate_config
+                    .advice_cycle_assigner()
+                    .assign_all_advice(&mut region, || "z0_primary", self.input.z_0.iter().copied())
+                    .map(|inp| inp.try_into().unwrap())
+            },
+        )?;
+
+        let assigned_z_next = <Self as StepCircuit<ARITY, C::Base>>::synthesize_step(
+            self,
+            config,
+            layouter,
+            &assigned_z_i,
+        )?;
+
+        Ok(assigned_z_next.map(|cell| cell.value().unwrap().cloned().unwrap()))
+    }
+
+    fn synthesize_step_base_case(
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
+        public_params: &StepParams<C::Base, RO>,
+        u: &PlonkInstance<C>,
```

Comment:
instead of `PlonkInstance`, use `AssignedPlonkInstance` here.

Reply:
https://github.com/snarkify/sirius/issues/153